### PR TITLE
Cancel Pending Host Applications when Host is Deactivated

### PR DESCRIPTION
**Pull request by SudharakaP**
Resolve https://github.com/opencollective/opencollective/issues/6278

[cancel-hosting-applications.webm](https://user-images.githubusercontent.com/12435965/214723391-4dd6dff0-b8d6-4d3a-8121-b849c23ab158.webm)

### CODE REVIEW

**Review by Betree — COMMENTED** [review]
Should we rather resolve the root cause for https://github.com/opencollective/opencollective/issues/6278? It seems legit to me to disable a host with pending applications, they should just be cancelled

File: server/models/Collective.js [hunk 1]
```diff
@@ -1207,6 +1208,19 @@ Collective.prototype.deactivateAsHost = async function () {
     );
   }
 
+  // Make sure we clean up all pending applications
+  await models.HostApplication.update(
+    { status: HostApplicationStatus.EXPIRED },
+    { where: { HostCollectiveId: this.id } },
+  );
+
+  await models.Member.update({ deletedAt: new Date() }, { where: { MemberCollectiveId: this.id, role: 'HOST' } });
```
Comment (Betree) [inline server/models/Collective.js:1217]: Do we really need that one? `HOST` member are not supposed to be created until the application gets approved

**Comment by Betree** [comment]
The overall approach looks good!

File: server/models/Collective.js (hunk 1 above)
Comment (SudharakaP) [reply server/models/Collective.js:1268]: @Betree : I've double checked this and it seems that we do add an entry to the `Member` table upon applying for the host; 

https://github.com/opencollective/opencollective-api/blob/b41f19b6509d12289772f39b9c1b1b3bf6f64058/server/models/Collective.js#L2343-L2348

**Comment by SudharakaP** [comment]
@Betree : Anything to do here? Okay to merge?

File: server/models/Collective.js [hunk 2]
```diff
@@ -1258,6 +1259,19 @@ Collective.prototype.deactivateAsHost = async function () {
...
+
+  await models.Collective.update(
+    { HostCollectiveId: null },
+    { hooks: false, where: { HostCollectiveId: this.id, isActive: false } },
+  );
+
```
Comment (Betree) [inline server/models/Collective.js:1274]: Not needed, we do this already in deactivateBudget: https://github.com/opencollective/opencollective-api/blob/dc25778c2ae332f27891f632d7fd09fc0ba9f7c1/server/models/Collective.js#L1410

```suggestion
```

**Review by Betree — COMMENTED** [review]
Can we add a small test in https://github.com/opencollective/opencollective-api/blob/dc25778c2ae332f27891f632d7fd09fc0ba9f7c1/test/server/models/Collective.test.js#L363, just to make sure we cover these new changes?

File: server/models/Collective.js (hunk 1 above)
Comment (Betree) [reply server/models/Collective.js:1268]: Interesting :+1:

**Comment by SudharakaP** [comment]
> Can we add a small test in
> 
> https://github.com/opencollective/opencollective-api/blob/dc25778c2ae332f27891f632d7fd09fc0ba9f7c1/test/server/models/Collective.test.js#L363
> , just to make sure we cover these new changes?

@Betree : I've added a test for checking this case. 👍🏽

File: server/models/Collective.js [hunk 3]
```diff
@@ -1261,6 +1262,19 @@ Collective.prototype.deactivateAsHost = async function () {
     );
   }
 
+  // Make sure we clean up all pending applications
+  await models.HostApplication.update(
+    { status: HostApplicationStatus.EXPIRED },
+    { where: { HostCollectiveId: this.id } },
+  );
+
+  await models.Member.update({ deletedAt: new Date() }, { where: { MemberCollectiveId: this.id, role: 'HOST' } });
```
Comment (Betree) [inline server/models/Collective.js:1271]: :ice_cream: Not a requested change, just a cleaner option IMO

```suggestion
  await models.Member.destroy({ where: { MemberCollectiveId: this.id, role: 'HOST' } });
```

File: test/server/models/Collective.test.js [hunk 4]
```diff
@@ -369,6 +369,38 @@ describe('server/models/Collective', () => {
...
+    it('auto cancel pending host applications when host is deactivated', async () => {
+      const temporaryHost = await models.Collective.create({
+        name: 'Temporary Host',
+        slug: 'temporaryHost',
+        hostFeePercent: 0,
+        type: 'ORGANIZATION',
+        currency: 'EUR',
+        CreatedByUserId: user1.id,
+        isHostAccount: true,
+      });
```
Comment (Betree) [inline test/server/models/Collective.test.js:381]: :ice_cream: Not a requested change: You can use the `fake-data` library for cleaner/simpler tests. Ex. here:

```suggestion
      const temporaryHost = await fakeHot({ admin: user1 });
```

**Review by Betree — APPROVED** [review]
Lgtm

File: test/server/models/Collective.test.js (hunk 4 above)
Comment (SudharakaP) [reply test/server/models/Collective.test.js]: Thanks. Done. 👍🏽

**Pull request merged**